src/tests/unit/tests/electron/platform/android/android-setup-telemetry-sender.test.ts

```diff
@@ -0,0 +1,111 @@
+// Licensed under the MIT License.
+import { TelemetryEventHandler } from 'background/telemetry/telemetry-event-handler';
+import { TelemetryEventSource, TriggeredByNotApplicable } from 'common/extension-telemetry-events';
+import {
+    AndroidSetupStepTelemetryData,
+    DEVICE_SETUP_STEP,
+} from 'electron/common/electron-telemetry-events';
+import { AndroidSetupStore } from 'electron/flux/store/android-setup-store';
+import { AndroidSetupTelemetrySender } from 'electron/platform/android/android-setup-telemetry-sender';
+import { AndroidSetupStepId } from 'electron/platform/android/setup/android-setup-step-id';
+import { It, Mock, MockBehavior, Times } from 'typemoq';
+
+describe('AndroidSetupTelemetrySender', () => {
+    it('sends appropriate telemetry events when steps change', () => {
+        testTelemetryEvents(
+            ['detect-adb', 'detect-devices', 'detect-permissions'],
+            [20, 400, 415],
+            [
+                {
+                    prevStep: null,
+                    newStep: 'detect-adb',
+                    prevDuration: 0,
+                },
+                {
+                    prevStep: 'detect-adb',
+                    newStep: 'detect-devices',
+                    prevDuration: 380,
+                },
+                {
+                    prevStep: 'detect-devices',
+                    newStep: 'detect-permissions',
+                    prevDuration: 15,
+                },
+            ],
+        );
+    });
+
+    it('does not send event if step has not changed', () => {
+        testTelemetryEvents(
+            ['detect-adb', 'detect-adb', 'detect-devices'],
+            [20, 400, 450],
+            [
+                {
+                    prevStep: null,
+                    newStep: 'detect-adb',
+                    prevDuration: 0,
+                },
+                {
+                    prevStep: 'detect-adb',
+                    newStep: 'detect-devices',
+                    prevDuration: 430,
+                },
+            ],
+        );
+    });
+
+    function testTelemetryEvents(
+        steps: AndroidSetupStepId[],
+        timestamps: number[],
+        expectedEvents: Omit<AndroidSetupStepTelemetryData, 'triggeredBy' | 'source'>[],
+    ): void {
+        const getCurrentMsMock = Mock.ofInstance(() => 0, MockBehavior.Strict);
+        const telemetryEventHandlerMock = Mock.ofType(TelemetryEventHandler, MockBehavior.Strict);
+        const androidSetupStoreMock = Mock.ofType(AndroidSetupStore, MockBehavior.Strict);
+        const testListener = new AndroidSetupTelemetrySender(
+            androidSetupStoreMock.object,
+            telemetryEventHandlerMock.object,
+            getCurrentMsMock.object,
+        );
+
+        steps.forEach(currentStepId =>
+            androidSetupStoreMock
+                .setup(m => m.getState())
+                .returns(_ => ({ currentStepId }))
+                .verifiable(Times.exactly(steps.length)),
```

Comment:
this seemed odd to me - it seems that the default (Times.once()) doesn't 'stack' like the return record-and-replay behavior does. This [codepen](https://codepen.io/karanbirsingh/pen/dyGRMJG?editors=0010) shows a repro. If the arguments in setup() are different, or the mock isn't strict, this goes away

Reply:
This looks like the same issue as https://github.com/florinn/typemoq/issues/94, but it kind of looks like typemoq is a dead project, so it seems unlikely that it's worth following up on :(